src/common/libkvs/treeobj.c

```diff
@@ -0,0 +1,496 @@
+        }
+    }
+    else if (!strcmp (type, "dir")) {
+        const char *key;
+        if (!json_is_object (data))
+            goto inval;
+        json_object_foreach (data, key, o) {
+            if (treeobj_validate (o) < 0)
+                goto inval;
+        }
+    }
+    else if (!strcmp (type, "symlink")) {
+        if (!json_is_string (data))
+            goto inval;
+    }
+    else if (!strcmp (type, "val")) {
+    }
+    else
+        goto inval;
+    return 0;
+inval:
+    errno = EINVAL;
+    return -1;
+}
+
+const char *treeobj_get_type (json_t *obj)
+{
+    const char *type;
+    if (treeobj_unpack (obj, &type, NULL) < 0)
+        return NULL;
+    return type;
+}
+
+bool treeobj_is_symlink (json_t *obj)
+{
+    const char *type = treeobj_get_type (obj);
+    return type && !strcmp (type, "symlink");
+}
+
+bool treeobj_is_val (json_t *obj)
+{
+    const char *type = treeobj_get_type (obj);
+    return type && !strcmp (type, "val");
+}
+
+bool treeobj_is_valref (json_t *obj)
+{
+    const char *type = treeobj_get_type (obj);
+    return type && !strcmp (type, "valref");
+}
+
+bool treeobj_is_dir (json_t *obj)
+{
+    const char *type = treeobj_get_type (obj);
+    return type && !strcmp (type, "dir");
+}
+
+bool treeobj_is_dirref (json_t *obj)
+{
+    const char *type = treeobj_get_type (obj);
+    return type && !strcmp (type, "dirref");
+}
+
+json_t *treeobj_get_data (json_t *obj)
+{
+    json_t *data;
+    if (treeobj_unpack (obj, NULL, &data) < 0)
+        return NULL;
+    return data;
+}
+
+int treeobj_decode_val (json_t *obj, void **dp, int *lp)
+{
+    int version;
+    const char *type, *xdata;
+    int len, xlen;
+    char *data;
+
+    if (json_unpack (obj, "{s:i s:s s:s}", "ver", &version,
+                                           "type", &type,
+                                           "data", &xdata) < 0
+                        || version != 1 || strcmp (type, "val") != 0) {
+        errno = EINVAL;
+        return -1;
+    }
+    xlen = strlen (xdata);
+    len = base64_decode_length (xlen);
+    if (len > 1) {
+        if (!(data = malloc (len))) {
+            errno = ENOMEM;
+            return -1;
+        }
+        if (base64_decode_block (data, &len, xdata, xlen) < 0) {
+            free (data);
+            errno = EINVAL;
+            return -1;
+        }
+    }
+    else {
+        len = 0;
+        data = NULL;
+    }
+    if (lp)
+        *lp = len;
+    if (dp)
+        *dp = data;
+    else
+        free (data);
+    return 0;
+}
+
+int treeobj_get_count (json_t *obj)
+{
+    json_t *data;
+    const char *type;
+    int count = -1;
+
+    if (treeobj_unpack (obj, &type, &data) < 0)
+        goto done;
+    if (!strcmp (type, "valref") || !strcmp (type, "dirref")) {
+        count = json_array_size (data);
+    }
+    else if (!strcmp (type, "dir")) {
+        count = json_object_size (data);
+    }
+    else if (!strcmp (type, "symlink") || !strcmp (type, "value")) {
+        count = 1;
```

Comment:
should be ```"val"```, not ```"value"```.  A unit test should have caught this, so perhaps a unit test is missing.